ggml/src/ggml.c

```diff
@@ -18833,63 +19370,43 @@ enum ggml_status ggml_graph_compute(struct ggml_cgraph * cgraph, struct ggml_cpl
             {
                 // update the number of threads from the actual number of threads that we got from OpenMP
                 n_threads = omp_get_num_threads();
-                state_shared.n_threads = n_threads;
+                threadpool->n_threads_cur = n_threads;
             }
 
-            struct ggml_compute_state worker = {
-                .thrd   = 0,
-                .ith    = omp_get_thread_num(),
-                .shared = &state_shared,
-            };
-            ggml_graph_compute_thread(&worker);
+            ggml_graph_compute_thread(&threadpool->workers[omp_get_thread_num()]);
         }
     } else {
-        struct ggml_compute_state worker = {
-            .thrd   = 0,
-            .ith    = 0,
-            .shared = &state_shared,
-        };
-        ggml_graph_compute_thread(&worker);
+        ggml_graph_compute_thread(&threadpool->workers[0]);
     }
 #else
-    struct ggml_compute_state * workers = alloca(sizeof(struct ggml_compute_state)*n_threads);
-
-    for (int j = 0; j < n_threads; ++j) {
-        workers[j] = (struct ggml_compute_state) {
-            .thrd   = 0,
-            .ith    = j,
-            .shared = &state_shared,
-        };
-    }
-
-    // create thread pool
-    for (int j = 1; j < n_threads; ++j) {
-        const int rc = ggml_thread_create(&workers[j].thrd, NULL, ggml_graph_compute_thread, &workers[j]);
-        GGML_ASSERT(rc == 0);
-        UNUSED(rc);
+    // Update main thread prio and affinity to match the current threadpool
+    ggml_thread_apply_process_priority(threadpool->prio);
+    ggml_thread_apply_thread_priority(threadpool->prio);
```

Comment:
Changing the thread priority is fine since the threads are owned by ggml, but I don't think that ggml should be changing the process priority. That should be entirely up to the application.

Reply:
Actually, on Windows we have to set overall Process Priority class in order to change Thread Priority.
https://learn.microsoft.com/en-us/windows/win32/procthread/scheduling-priorities

We can collapse these calls into one though and only update process priority class on Windows.
Of course, if the app doesn't have any other threads of its own then changing main thread priority by definition means changing the process. 

btw One of the changes @fmz and I wanted to propose (not right now but at some point once the core bits are merged) is to do all of the `graph_schedule/compute()` work in the threadpool workers. i.e don't use the main thread for any actual work. 
That is a better scheme for not interfering with the app thread(s) where ggml happened to be call from.
The app may want to run on the efficiency cores so its main thread (might not even be main) might not be usable.
This thread should just set things up and then dispatch work to the threadpool and wait on cond.var for completion.

Reply:
I understand that thread priorities have different semantics in windows and linux, but I still don't think that's a good reason to change the process priority. If anything, the windows behavior seems much more sane to me. If the application actually wants to change the process priority, they can easily do it themselves. For example, I would expect that a common use case for this would be to set the compute threads to a lower priority to avoid interfering with the application/UI threads, and that doesn't require changing the process priority (if anything, lowering the process priority would be undesirable).


> btw One of the changes @fmz and I wanted to propose (not right now but at some point once the core bits are merged) is to do all of the `graph_schedule/compute()` work in the threadpool workers. i.e don't use the main thread for any actual work.

I think it would be good to have the option to run all the computation asynchronously in secondary threads. I discussed this briefly in https://github.com/ggerganov/ggml/issues/721.

Reply:
Ok. I'll nuke the process priority changes from within thre threads and add it to the `llama-cli` and `llama-bench` instead. I'll reintroduce `ggml_apply_process_priority()` to support simple mapping that I added (ie into OS specific priority class), and will call it only from the apps if --prio is specified. 

And sounds good on moving the computation to secondary threads. Will play with that in the threadpool V3.